app/src/androidTest/java/org/simple/clinic/storage/migrations/BaseDatabaseMigrationTest.kt

```diff
@@ -0,0 +1,37 @@
+package org.simple.clinic.storage.migrations
+
+import androidx.annotation.CallSuper
+import androidx.room.migration.Migration
+import androidx.sqlite.db.SupportSQLiteDatabase
+import org.junit.Before
+import org.junit.Rule
+import org.junit.rules.ExpectedException
+import org.simple.clinic.MigrationTestHelperWithForeignKeyConstraints
+import org.simple.clinic.TestClinicApp
+import javax.inject.Inject
+
+abstract class BaseDatabaseMigrationTest(
+    private val fromVersion: Int,
+    private val toVersion: Int
+) {
+
+  @get:Rule
+  val helper = MigrationTestHelperWithForeignKeyConstraints()
+
+  @get:Rule
+  val expectedException: ExpectedException = ExpectedException.none()
+
+  @Inject
+  lateinit var migrations: List<@JvmSuppressWildcards Migration>
+
+  @Before
+  @CallSuper
+  open fun setUp() {
+    TestClinicApp.appComponent().inject(this)
+    helper.migrations = migrations
+  }
+
+  protected val before: SupportSQLiteDatabase by lazy(LazyThreadSafetyMode.NONE) { helper.createDatabase(fromVersion) }
+
+  protected val after: SupportSQLiteDatabase by lazy(LazyThreadSafetyMode.NONE) { helper.migrateTo(toVersion) }
```

Comment:
In my opinion, abstracting `before` and `after` in tests adds a bit of cognitive complexity in understanding the migrations tests themselves. We will have to go and look what `before` and `after` means. Also that would abstract the `given` and `when` while writing a test. Specially the `given` when we are not performing assertions on it.

Reply:
I find this to be quite neatly done. One would have to look at the source probably only once to understand what `before` and `after` is as they are just one-line definitions. So IMO, we can leave this as is.
I don't think I fully understand what the "abstraction of `given` and `when`" implies.

Reply:
@rakshakhegde To answer your [question](https://github.com/simpledotorg/simple-android/pull/979#discussion_r359188926):
We follow the `given` and `when` convention to write tests.
In our migration tests, the `given` is what `before` does here and `when` is what `after` does. We will have to read `before` and `after` every time we write these tests. That is what I mean by "abstract the given and when".

Reply:
But if you think that this is fine, then I will keep this as is. It was a comment for @vinaysshenoy to see later.

Reply:
My intention was to hide the details of the database migration itself. If you look at how we currently write a migration test, it goes something like:

```kotlin
val db_v1 = /* setup database */

val db_v2 = /* migrate db_v1 */

// verify db_v2
```

It is not actually important to know which version is being migrated to which. What is important is the state of the database **before** migration and the state of the database **after** migration. This was the reason I chose to name them `before` and `after` so that we can focus on the migration itself instead of anything else.

I felt that it is not necessary to know what `before` and `after` actually do, since they are quite evident. Since this test is written specifically to test database migrations, someone can just use `before` and `after` to setup and verify the migrations without needing to write the boilerplate to do the migrations themselves.

Does this make sense?